libs/s25main/desktops/dskOptions.cpp

```diff
@@ -663,3 +688,91 @@ void dskOptions::loadVideoModes()
     // Sort by aspect ratio
     std::sort(video_modes.begin(), video_modes.end(), cmpVideoModes);
 }
+
+void dskOptions::Msg_ScreenResize(const ScreenResizeEvent& sr)
+{
+    Desktop::Msg_ScreenResize(sr);
+    updateGuiScale();
+}
+
+bool dskOptions::Msg_WheelUp(const MouseCoords& mc)
+{
+    if(!scrollGuiScale(true))
+        return Desktop::Msg_WheelUp(mc);
+    return true;
+}
+
+bool dskOptions::Msg_WheelDown(const MouseCoords& mc)
+{
+    if(!scrollGuiScale(false))
+        return Desktop::Msg_WheelDown(mc);
+    return true;
+}
+
+void dskOptions::updateGuiScale(ctrlComboBox* combo)
+{
+    // generate GUI scale percentages in 10% increments
+    constexpr auto stepSize = 10u;
+    const auto roundGuiScale = [=](unsigned percent) { return (percent + stepSize / 2) / stepSize * stepSize; };
+
+    unsigned minPercent, maxPercent, recommendedPercent;
+    VIDEODRIVER.getGuiScaleRange(&minPercent, &maxPercent, &recommendedPercent);
+    const auto recommendedPercentRounded = roundGuiScale(recommendedPercent);
+
+    if(!combo)
+    {
+        auto* group = GetCtrl<ctrlGroup>(ID_grpGraphics);
+        combo = group->GetCtrl<ctrlComboBox>(ID_cbGuiScale);
+    }
+
+    guiScales_.clear();
+    combo->DeleteAllItems();
+
+    combo->AddString((boost::format(_("Auto (%u%%)")) % recommendedPercent).str());
+    if(SETTINGS.video.guiScale == 0)
+        combo->SetSelection(0);
+
+    for(unsigned percent = roundGuiScale(minPercent); percent <= maxPercent; percent += stepSize)
+    {
+        if(percent == recommendedPercentRounded)
+            recommendedGuiScaleIndex_ = guiScales_.size();
+        guiScales_.push_back(percent);
+
+        combo->AddString(helpers::toString(percent) + "%");
+        if(percent == SETTINGS.video.guiScale)
+            combo->SetSelection(combo->GetNumItems() - 1);
+    }
+
+    // if GUI scale exceeds maximum, lower it to keep UI elements on screen
+    if(SETTINGS.video.guiScale > guiScales_.back())
+    {
+        combo->SetSelection(combo->GetNumItems() - 1);
+        SETTINGS.video.guiScale = guiScales_.back();
+        VIDEODRIVER.setGuiScale(SETTINGS.video.guiScale);
+    }
+}
+
+bool dskOptions::scrollGuiScale(bool up)
+{
+    if(!VIDEODRIVER.GetModKeyState().ctrl)
+        return false;
+
+    unsigned index = 0;
+    auto* group = GetCtrl<ctrlGroup>(ID_grpGraphics);
+    auto* combo = group->GetCtrl<ctrlComboBox>(ID_cbGuiScale);
+    const auto& selection = combo->GetSelection();
+    if(!selection || *selection == 0) // No selection or "Auto" item selected
+        index = recommendedGuiScaleIndex_ + 1;
```

Comment:
This ignores the "up" parameter AND the range of `guiScales_`. Idea:   
Set `index` to `selection` or `recommendedGuiScaleIndex_`, then check `up` and range using `index` instead of `selection`.

Reply:
I'll check. I fiddled around with these lines quite a bit trying to make them easier to read while still retaining the desired behavior.

Reply:
> This ignores the "up" parameter AND the range of `guiScales_`.

I didn't quite understand what you meant by this earlier. No, this is deliberate. Regardless of direction, we start at the recommended index. The `+ 1` accounts for "Auto" in the combo box. Note that "Auto" and the recommended percentage in the combo box aren't required to be the same due to rounding. Hence, storing the index instead of using the value obtained via `GetGuiScaleRange()`.

Nevertheless, I'll see if I can improve clarity.

Reply:
I guess using an extra variable to hold the "old selection" (taking the unset & auto case into account) and only checking `up` together with `std::min/max` might work. Just a suggestion.   
Important to me is that this case is handled correctly, i.e. not choosing a higher scale when scrolling down

Reply:
"Auto" and none always move to the recommended index first. That means that given a recommended scale of 152%, moving in either direction would select 150% first. If we want "Auto" to move to 140% and 160% respectively, which would make sense, we can make the logic more complicated. ;-)

I'll give it a shot later and report back.

(FYI, `std::clamp` was added in C++17. Yay!)

Reply:
> "Auto" and none always move to the recommended index first. That means that given a recommended scale of 152%, moving in either direction would select 150% first. If we want "Auto" to move to 140% and 160% respectively, which would make sense, we can make the logic more complicated. ;-)

The first thing makes sense to me: Moving in either directly selects the recommended index first, i.e. effectively a no-op. So `        index = recommendedGuiScaleIndex_ + 1;` is a bit misleading because `index` is not the "index" but the new selection

So your approach makes sense to me, just the naming `index = index-we-want + 1` threw me off.

> (FYI, std::clamp was added in C++17. Yay!)

Oh right! IIRC we even had a replacement in `helpers`-namespace.

Reply:
> Oh right! IIRC we even had a replacement in `helpers`-namespace.

I think so too.